firedrake/dmplex.pyx

```diff
@@ -1014,3 +964,784 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+    raise RuntimeError("This should never happen!")
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef locally_orient_quadrilateral_plex(PETSc.DM plex,
+                                       PETSc.Section vertex_numbering,
+                                       np.int32_t *cell_ranks,
+                                       np.int32_t *facet2index,
+                                       np.int32_t nfacets_shared,
+                                       np.int8_t *orientations):
+    """Locally orient the facets (edges) of a quadrilateral plex, and
+    derive the dependency information of shared facets.
+
+    :arg plex: The DMPlex object encapsulating the mesh topology
+    :arg vertex_numbering: Section describing the universal vertex numbering
+    :arg cell_ranks: MPI rank of the owner of each (visible) non-owned cell,
+                     or -1 for (locally) owned cell.
+    :arg facet2index: Maps plex facet numbers to their index in the buffer
+                      of shared facets.
+    :arg nfacets_shared: Number of facets shared with other MPI nodes.
+    :arg orientations: Facet orientations relative to the plex.
+                       -1: orientation not set
+                        0: orientation is same as in (local) plex
+                       +1: orientation is the opposite of that in plex
+                       THIS ARRAY IS UPDATED BY THIS FUNCTION.
+
+    Returns an array of size 'nfacets_shared', which tells for each shared
+    facet which other shared facet needs update, if any, when a shared facet
+    is flipped.
+     * Equal to 'nfacets_shared': no other facet requires update.
+     * Non-negative value: index of shared facet,
+                           which must have the same global orientation.
+     * Negative value 'i': cneg(i) is the index of the shared facet,
+                           which must have opposite global orientation.
+    """
+    cdef:
+        PetscInt nfacets, fStart, fEnd, f
+        PetscInt size
+        PetscInt support[2]
+        PetscInt start_facet, end_facet
+        np.int8_t twist
+        PetscInt i, j
+        np.ndarray[np.int32_t, ndim=1, mode="c"] result
+
+    fStart, fEnd = plex.getHeightStratum(1)
+    nfacets = fEnd - fStart
+
+    result = np.empty(nfacets_shared, dtype=np.int32)
+
+    for f in range(fStart, fEnd):
+        if orientations[f - fStart] < 0:
+            orientations[f - fStart] = 0
+
+            plex_get_restricted_support(plex, cell_ranks, f, &size, support)
+            assert 0 <= size <= 2
+            if size == 0:
+                continue
+
+            end_facet = traverse_cell_string(plex, f, support[0],
+                                             cell_ranks, orientations)
+            if end_facet == -1:
+                # Closed loop
+                if orientations[f - fStart]:
+                    # Moebius strip found
+                    raise RuntimeError("Moebius strip found in the mesh.")
```

Comment:
OK, so if we came round a loop and found that the last cell we hit claims that the end_facet must be flipped then there must be a twist in the loop, because the start_facet (which is the same) should have had no flip.